### ISSUE TITLE
Deserialization error in NMF

### ISSUE THREAD

**Issue by npyoung** [thunder-project/thunder#129]
```python
example_data = tsc.loadExample('fish-series')
nmf = thunder.NMF(k=5, method='als')
nmf.fit(example_data)
```
Gives me the following error:
```python
<ipython-input-18-4c8d4d054a3d> in <module>()
      1 example_data = tsc.loadExample('fish-series')
      2 nmf = thunder.NMF(k=5, method='als')
----> 3 nmf.fit(example_data)

/mnt/spark/spark-1841c1f0-1813-49ed-ab38-36a1bb4eabd0/spark-eb4f748e-97a3-423c-81fd-326f1c953cc6/thunder_python-0.5.0_dev-py2.6.egg/thunder/factorization/nmf.py in fit(self, mat)
    160 
    161                 # update H using least squares row-wise with inv(W' * W) * W * R (same as pinv(W) * R)
--> 162                 h = pinvW.values().zip(mat.values()).map(lambda (x, y): outer(x, y)).reduce(add)
    163 
    164                 # clip negative values of H

/root/spark/python/pyspark/rdd.pyc in reduce(self, f)
    723             yield reduce(f, iterator, initial)
    724 
--> 725         vals = self.mapPartitions(func).collect()
    726         if vals:
    727             return reduce(f, vals)

/root/spark/python/pyspark/rdd.pyc in collect(self)
    684         """
    685         with SCCallSiteSync(self.context) as css:
--> 686             bytesInJava = self._jrdd.collect().iterator()
    687         return list(self._collect_iterator_through_file(bytesInJava))
    688 

/root/spark/python/lib/py4j-0.8.2.1-src.zip/py4j/java_gateway.py in __call__(self, *args)
    536         answer = self.gateway_client.send_command(command)
    537         return_value = get_return_value(answer, self.gateway_client,
--> 538                 self.target_id, self.name)
    539 
    540         for temp_arg in temp_args:

/root/spark/python/lib/py4j-0.8.2.1-src.zip/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
    298                 raise Py4JJavaError(
    299                     'An error occurred while calling {0}{1}{2}.\n'.
--> 300                     format(target_id, '.', name), value)
    301             else:
    302                 raise Py4JError(

Py4JJavaError: An error occurred while calling o355.collect.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 24.0 failed 4 times, most recent failure: Lost task 0.3 in stage 24.0 (TID 39, ip-10-168-156-186.ec2.internal): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/root/spark/python/pyspark/worker.py", line 107, in main
    process()
  File "/root/spark/python/pyspark/worker.py", line 98, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/root/spark/python/pyspark/serializers.py", line 231, in dump_stream
    vs = list(itertools.islice(iterator, batch))
  File "/root/spark/python/pyspark/rdd.py", line 723, in func
    yield reduce(f, iterator, initial)
  File "/root/spark/python/pyspark/serializers.py", line 301, in load_stream
    " in pair: (%d, %d)" % (len(keys), len(vals)))
ValueError: Can not deserialize RDD with different number of items in pair: (128, 64)
```
I haven't run the NMF unit test, but I think it might be missing this because it only uses 1D keys.

Using 2af3738 and spark 1.2.1

**Comment by freeman-lab** [comment]
Thanks for reporting @npyoung , weirdly, this appears to be specific to Spark 1.2.1! I'm reproducing it with that version, but not with Spark 1.2.0 (same machine, same thunder version). I don't understand the problem yet, but I suspect it's specific to the zip operation, because other matrix related operations on the same data work fine (PCA, ICA). I'm also finding that the unit tests run fine, so it could be related to the keys as you suggest.

**Comment by freeman-lab** [comment]
Closing, as we found this to be an as-yet-unresolved Spark error specific to 1.2.1, fixed as of Spark 1.3, but not apparently related to Thunder itself.
